### PR TITLE
Make SFT hardware-agnostic

### PR DESCRIPTION
**Pull request by DamianSzwichtenberg**
This PR enables the SFT (Supervised Fine-Tuning) application to work on XPU hardware by abstracting device management from CUDA-specific APIs to hardware-agnostic PyTorch accelerator APIs.

## Changes:
- Replaced CUDA-specific API calls (torch.cuda.*) with hardware-agnostic accelerator APIs (torch.accelerator.*) throughout the codebase
- Introduced a DeviceProxy class to handle device counting and environment variable mapping for different hardware backends (CUDA, XPU)
- Updated test files to use generic device visibility terminology and mock accelerator APIs instead of CUDA-specific mocks

### CODE REVIEW

**Comment by felipemello1** [comment]
thanks for the PR @DamianSzwichtenberg , this is very useful. We are on an offsite this week, can i get back to you in a few days?

**Comment by DamianSzwichtenberg** [comment]
@felipemello1 No problem, enjoy!

**Comment by DamianSzwichtenberg** [comment]
@felipemello1 Just a gentle bump on this.

File: src/forge/controller/provisioner.py [hunk 1]
```diff
@@ -356,7 +406,9 @@ async def get_proc_mesh(
                 # Set the PTD world size
                 world_size = num_procs * (num_hosts or 1)
                 env_vars["WORLD_SIZE"] = str(world_size)
-                env_vars["CUDA_VISIBLE_DEVICES"] = ",".join(gpu_ids)
+
+                # Set device isolation using the appropriate environment variable
+                env_vars.update(DeviceProxy.get_isolation_env_vars(gpu_ids))
```
Comment (felipemello1) [inline src/forge/controller/provisioner.py:411]: how sure are we that the key here is correct, since we dont test the key anymore? 

Image

File: src/forge/observability/perf_tracker.py [hunk 2]
```diff
@@ -272,36 +272,36 @@ class _TimerCUDA(_TimerProtocol):
...
-            raise RuntimeError("CUDA is not available for timing")
+        if not torch.accelerator.is_available():
+            raise RuntimeError("Accelerator is not available for timing")
         self._executor = ThreadPoolExecutor(max_workers=max_workers)
         self._futures: list[tuple[str, Future[float], int]] = (
             []
         )  # (name, future, submission_index)
         self._durations: list[tuple[str, float]] = []
-        self._chain_start: torch.cuda.Event | None = None
+        self._chain_start: torch.Event | None = None
```
Comment (felipemello1) [inline src/forge/observability/perf_tracker.py:282]: not familiar with these APIs, is this correct? Just double checking

File: src/forge/observability/perf_tracker.py [hunk 3]
```diff
@@ -185,23 +185,23 @@ def _start_memory_tracking(self) -> None:
 
         if should_track:
             _set_memory_active(True)
-            torch.cuda.reset_peak_memory_stats()
-            self._start_mem = torch.cuda.memory_allocated()
+            torch.accelerator.reset_peak_memory_stats()
+            self._start_mem = torch.accelerator.memory_allocated()
```
Comment (felipemello1) [inline src/forge/observability/perf_tracker.py:189]: curious to see an wandb with main vs this branch and compare results to make sure they are the same

**Review by felipemello1 — COMMENTED** [review]
at first glance the PR looks great! The provisioner looks much cleaner. 

I would feel more confident merging it if we had an wandb comparing main vs this branch for the same setup, anything with >1 gpu. Would you have the resources to provide that, if its not asking too much? (you can say no)

I also feel like i should have claude take a look at this and see if we missed some important .cuda vs .accelerator, and if the test changes are not "masking" any errors. I will get this done this week.

thanks for contributing with this PR!

File: src/forge/controller/provisioner.py (hunk 1 above)
Comment (DamianSzwichtenberg) [reply src/forge/controller/provisioner.py:411]: The key is still tested, just indirectly. Here's the chain:
- the autouse fixture mocks [current accelerator](https://github.com/meta-pytorch/torchforge/pull/749/changes#diff-680c5d6d87a3b5c6f91241e86adf7504aae546837a2c2216845acd7e9fcc4524R24) to return cuda
- [mock_patch_visible_devices_var](https://github.com/meta-pytorch/torchforge/pull/749/changes#diff-680c5d6d87a3b5c6f91241e86adf7504aae546837a2c2216845acd7e9fcc4524R16) sets CUDA_VISIBLE_DEVICES

File: src/forge/observability/perf_tracker.py (hunk 2 above)
Comment (DamianSzwichtenberg) [reply src/forge/observability/perf_tracker.py:282]: Yes, `Event` will query the current accelerator type.

File: src/forge/observability/perf_tracker.py (hunk 3 above)
Comment (DamianSzwichtenberg) [reply src/forge/observability/perf_tracker.py:189]: Here you go: [link](https://api.wandb.ai/links/damian-szwichtenberg-intel/lgtm4ix0)
